File: Backend/DotNet/CredMann/CredMann/Crypto/Password/PasswordStrengthener.cs
using CredMann.Common;
using CredMann.Utils;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Prng;
using System;

namespace CredMann.Crypto.Password
{
    public class PasswordStrengthener
    {
        private PbeParametersGenerator parametersGenerator;
        private DigestRandomGenerator digestRandomGenerator;

        public PasswordStrengthener(DigestRandomGenerator digestGenerator, PbeParametersGenerator paramGenerator)
        {
            parametersGenerator = paramGenerator;
            digestRandomGenerator = digestGenerator;
        }

        #region Salt generators

        public byte[] GenerateSalt(int size = 16)
        {
            byte[] newSalt = new byte[size];
            digestRandomGenerator.NextBytes(newSalt);
            return newSalt;
        }

        #endregion

        #region Key generators
        public byte[] GenerateKey()
        {
            //Fixme pick from configuration
            return UtilFactory.Instance.GetRandomDataGenerator().GenerateKey(DefaultConfig.Default.KeySize);
        }

        public byte[] GenerateKey(ref string password, byte[] salt, int iterationCount = CommonConstants.DefaultIterationCount)
        {
            KeyParameter keyParam;
            char[] pass = null;

            try
            {
                if (string.IsNullOrWhiteSpace(password))
                    throw new ArgumentNullException(nameof(password));

                //Convert from securestring to regular char array
                pass = password.ToCharArray();

                keyParam = (KeyParameter)GenerateKeyParam(ref pass, ref salt, iterationCount);
            }
            finally
            {
                //Make sure the copies dont preserve any copy of the user's password
                if (pass != null)
                {
                    for (int i = 0; i < pass.Length; i++)
                        pass[0] = '0';
                }
            }

            return keyParam.GetKey();
        }

        public ParametersWithIV GenerateParametersWithIV(ref string password, byte[] salt, int iterationCount = CommonConstants.DefaultIterationCount)
        {
            ParametersWithIV paramWithIV;
            char[] pass = null;

            try
            {
                if (string.IsNullOrWhiteSpace(password))
                    throw new ArgumentNullException(nameof(password));

                //Convert from securestring to regular char array
                pass = password.ToCharArray();

                var keyParam = GenerateKeyParam(ref pass, ref salt, iterationCount);
                paramWithIV = new ParametersWithIV(keyParam, salt);
            }
            finally
            {
                //Make sure the copies dont preserve any copy of the user's password
                if (pass != null)
                {
                    for (int i = 0; i < pass.Length; i++)
                        pass[0] = '0';
                }
            }

            return paramWithIV;
        }

        public ParametersWithIV GenerateParametersWithIV(ref byte[] password, byte[] salt, int iterationCount = CommonConstants.DefaultIterationCount)
        {
            ParametersWithIV paramWithIV;
            char[] pass = null;

            try
            {
                if (password == null || password.Length == 0)
                    throw new ArgumentNullException(nameof(password));

                pass = DefaultConfig.Default.Encoding.GetString(password).ToCharArray();

                var keyParam = GenerateKeyParam(ref pass, ref salt, iterationCount);
                paramWithIV = new ParametersWithIV(keyParam, salt);
            }
            finally
            {
                //Make sure the copies dont preserve any copy of the user's password
                if (pass != null)
                {
                    for (int i = 0; i < pass.Length; i++)
                        pass[0] = '0';
                }
            }

            return paramWithIV;
        }

        #endregion

        #region Private members

        private ICipherParameters GenerateKeyParam(ref char[] password, ref byte[] salt, int iterationCount = CommonConstants.DefaultIterationCount)
        {
            //Initialize the generator
            parametersGenerator.Init(
                PbeParametersGenerator.Pkcs5PasswordToUtf8Bytes(password),
                salt,
                iterationCount);

            //Set algorithm and key size
            //RC4, AES            
            return parametersGenerator.GenerateDerivedParameters(CommonConstants.DerrivedParamsAlgorithm, DefaultConfig.Default.KeySize);
        }

        #endregion
    }
}

File: Backend/Java/JFxPasswordManagerTryout/src/main/java/com/spin/apps/password_manager/crypto/PasswordStrengthener.java
package com.spin.apps.password_manager.crypto;

import com.spin.apps.password_manager.AppConfiguration;
import org.bouncycastle.crypto.PBEParametersGenerator;
import org.bouncycastle.crypto.generators.PKCS5S2ParametersGenerator;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.stereotype.Component;

import javax.crypto.spec.SecretKeySpec;

@Component
@Configuration
public class PasswordStrengthener {

    @Autowired
    AppConfiguration configuration;

    @Autowired
    RandomDataBuilder randomDataBuilder;

    //Generator for PBKDF2 according to PKCS5-S2 specification
    final PBEParametersGenerator generator = new PKCS5S2ParametersGenerator();

    public byte[] generateSalt()
    {
        return randomDataBuilder.generateSalt(configuration.getLengthInBytes());
    }

    public byte[] generateKey()
    {
        return randomDataBuilder.generateRandomKey(configuration.getLengthInBytes());
    }

    public byte[] generate_Key(char[] password, byte[] salt)
    {
        return generate_Key(password, salt, 2000);
    }

    public byte[] generate_Key(char[] password, byte[] salt, int iterationCount)
    {
        generator.init(PBEParametersGenerator.PKCS5PasswordToUTF8Bytes(password), salt, iterationCount);
        KeyParameter parameter = (KeyParameter) generator.generateDerivedParameters(configuration.getLengthInBits());
        return parameter.getKey();
    }

    public ParametersWithIV generate_ParametersWithIV(char[] password, byte[] salt, int iterationCount)
    {
        generator.init(PBEParametersGenerator.PKCS5PasswordToUTF8Bytes(password), salt, iterationCount);
        return (ParametersWithIV) generator.generateDerivedParameters(configuration.getLengthInBits(), configuration.getLengthInBits());
    }
}
